clippy_lints/src/not_exhaustive_enough.rs

```diff
@@ -0,0 +1,172 @@
+use crate::utils::span_lint_and_sugg;
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::{Arm, Expr, ExprKind, FieldPat, MatchSource, Pat, PatKind, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:**
+    /// As suggested in the [non_exhaustive RFC](https://github.com/rust-lang/rfcs/blob/master/text/2008-non-exhaustive.md#unresolved-questions), when using non-exhaustive enums and structs in patterns,
+    /// the lint would warn the user for missing variants or fields despite having a wildcard arm or a rest pattern.
+    ///
+    /// **Why is this bad?**
+    /// When new fields/variants are added by the upstream crate they might go unnoticed.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust
+    /// #[non_exhaustive]
+    /// # enum E {First,Second,Third}
+    /// # let e = E::First;
+    /// // Bad
+    /// match e {
+    ///         E::First => {}
+    ///         E::Second => {}
+    ///         _ => {}
+    ///     }
+    /// // Good
+    /// match e {
+    ///         E::First => {}
+    ///         E::Second => {}
+    ///         E::Third => {}
+    ///     }
```

Comment:
The point of `#[non_exhaustive]` is that the wildcard match is always required:
```suggestion
    ///         E::Third => {}
    ///         _ => {}
    ///     }
```